comp/apm/etwtracer/impl/apmetwtracerimpl.go

```diff
@@ -0,0 +1,334 @@
+package apmetwtracerimpl
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"github.com/DataDog/datadog-agent/comp/apm/etwtracer"
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/comp/etw"
+	"github.com/DataDog/datadog-agent/pkg/util/fxutil"
+	etwutil "github.com/DataDog/datadog-agent/pkg/util/winutil/etw"
+	"github.com/Microsoft/go-winio"
+	"go.uber.org/fx"
+	"golang.org/x/sys/windows"
+	"io"
+	"net"
+	"time"
+	"unsafe"
+)
+
+// Module defines the fx options for this component.
+var Module = fxutil.Component(
+	fx.Provide(newApmEtwTracerImpl),
+)
+
+type dependencies struct {
+	fx.In
+	Lc  fx.Lifecycle
+	Log log.Component
+	Etw etw.Component
+}
+
+// pidContext wraps a named-pipe connection and a last-seen
+// timestamp for a given PID.
+type pidContext struct {
+	conn     net.Conn
+	lastSeen time.Time
+}
+
+type pidMap = map[uint64]pidContext
+
+func newApmEtwTracerImpl(deps dependencies) (apmetwtracer.Component, error) {
+	// Microsoft-Windows-DotNETRuntime
+	guid, _ := windows.GUIDFromString("{E13C0D23-CCBC-4E12-931B-D9CC2EEE27E4}")
+
+	apmEtwTracer := &apmetwtracerimpl{
+		pids:                      make(pidMap),
+		dotNetRuntimeProviderGUID: guid,
+		log:                       deps.Log,
+		etw:                       deps.Etw,
+		stopGarbageCollector:      make(chan struct{}),
+	}
+
+	// Cache the magic header
+	for idx := range magicHeaderString {
+		apmEtwTracer.magic[idx] = magicHeaderString[idx]
+	}
+
+	deps.Lc.Append(fx.Hook{OnStart: apmEtwTracer.start, OnStop: apmEtwTracer.stop})
+	return apmEtwTracer, nil
+}
+
+type apmetwtracerimpl struct {
+	session                   etw.Session
+	dotNetRuntimeProviderGUID windows.GUID
+
+	pids pidMap
+
+	pipeListener         net.Listener
+	log                  log.Component
+	etw                  etw.Component
+	magic                [14]byte
+	stopGarbageCollector chan struct{}
+}
+
+type header struct {
+	Magic           [14]byte
+	Size            uint16
+	CommandResponse uint8
+}
+
+type clrEvent struct {
+	header
+	EventHeader    etw.DDEventHeader
+	UserDataLength uint16
+	UserData       []byte
+}
+
+const (
+	magicHeaderString   = "DD_ETW_IPC_V1"
+	serverNamedPipePath = "\\\\.\\pipe\\DD_ETW_DISPATCHER"
+	clientNamedPipePath = "\\\\.\\pipe\\DD_ETW_CLIENT_%d"
+	headerSize          = 17
+	okResponse          = 0
+	registerCommand     = 1
+	unregisterCommand   = 2
+	clrEventResponse    = 16
+	errorResponse       = 255
+)
+
+type win32MessageBytePipe interface {
+	CloseWrite() error
+}
+
+func (a *apmetwtracerimpl) handleConnection(c net.Conn) {
+	// calls https://github.com/microsoft/go-winio/blob/e6aebd619a7278545b11188a0e21babea6b94182/pipe.go#L147
+	// which closes a pipe in message-mode
+	defer func(pipe win32MessageBytePipe) {
+		_ = pipe.CloseWrite()
+	}(c.(win32MessageBytePipe))
+
+	a.log.Debugf("client connected [%s]", c.RemoteAddr().Network())
+	for {
+		h := header{}
```

Comment:
> , but typically those read / write call would immediately return when the underlying named-pipe is closed, which happens when the Agent shutsdown.

Where the connection is closed outside of the function on shutdown.

Reply:
That is listener, I was thinking of closing this connection https://github.com/DataDog/datadog-agent/pull/20586/files#diff-02cea395ba46f80167e6df53ed4ed10724bdfb91577aed2793aa3984f7dc6620R259. Would pipe listener close affect spawn connection?

Reply:
I am passing review. I feel that you are right, because there is only one real pipe underneath. But it bothers me a bit existence of 2 objects (listener and connection itself). Can you for the hell of experiment to test with adding superficial code closure of a listener and seeing read or write on connection failed and loop is exited.